feat(post): modify endpoints to include Topics

wanlingt opened the pull request:
## Problem

Topics are not included in Post endpoints

Closes #399 

## Solution
Modify the following:
- `listPosts`
  - Add `topics` as a param to filter posts. More than 1 topic may be specified. Returns posts belonging to the specified topic(s) and their child topics.
  - Add `agencyId` as a param to filter posts. Validity checks for agency are already done on the client-side.
  - Remove merging of agency and tags; `agency` is no longer included in `tags` param
  - Refactor queries - only one query is used to retrieve posts. When more than 1 topic / tag is specified, posts belonging to those topics / tags are returned, instead of posts which match that exact combination of topics / tags
- `listAnswerablePosts`
  -  Refactor queries in the same manner as `listPosts`
  - Refactor splitting of tags, move splitting of tags from `post.routes` to `post.service`
- `getSinglePost`
  - Add `topic` as a param 
- `createPost`
  - Add `topicname` as a param 
  - Error thrown if no topics and tags are specified; at least one topic or tag is necessary to create a post
- `updatePost`
  - Add `topicname` as a param 
- `getRelatedPosts`
  - Look for posts belonging to the same topic, followed by posts with the most number of overlapping tags, followed by number of views

**Improvements**:

- Replace redundant `PostWithRelations` with `Posts`


## Before & After Screenshots
### For `getRelatedPosts`

**BEFORE**:
![image](https://user-images.githubusercontent.com/56983748/138433922-f54070cd-11bb-4e49-be9c-2fe1c971c919.png) 
**AFTER**:
<img width="1174" alt="Screenshot 2021-10-22 at 5 45 13 PM" src="https://user-images.githubusercontent.com/56983748/138434468-ce9f5ecb-ebb8-4083-ab1f-9696e22b965c.png">

## Tests
- Adding and removing tags from a post should work the same as before
- If multiple tags are specified when filtering posts by tags, posts returned should belong to at least one of the specified tags. Previously, posts returned had to have exact same combination of tags.

wanlingt committed:
feat(post): modify getSinglePost to include post's topic

- add topicId to PostWithRelations type \n- Change PostWithUserTagRelations to
PostWithUserTopicTagRelations \n- Change PostWithUserTagRelatedPostRelations to
PostWithUserTopicTagRelatedPostRelations \n- getSinglePost returns topicId in addition to existing
attributes

re #399

wanlingt committed:
feat(post): add topic to post creation

add topicname when creating post
- add topicId to post base model
- modify post tests

re #399

wanlingt committed:
fix(post): fix getSinglePost to return post with topicId

re #399

wanlingt committed:
refactor(post): include agency as param in listPosts

- add agency shortname as a param to filter posts
- refactor tag filtering so that only 1 request
is used

re #399

wanlingt committed:
feat(post): add topics to listPosts params

- add topics as a param in listPost for filtering
- topic filtering only occurs if a valid agency id
param is used
- modified query params in client QuestionsList component
- more than one topic can be used to filter posts

re #399

wanlingt committed:
test(post): modified post.controller error thrown for invalid topics

- added error message for invalid topics used in request
- modified test accordingly

re #399

wanlingt committed:
feat(topic): add topics to listAnswerablePosts params

- add topics as a param in listAnswerablePosts for filtering
- refactored query such that only one
query is used to obtain posts
- refactored splitting of tags, moved from routes to service
-
more than one topic can be used to filter posts

re #399

wanlingt committed:
feat(post): topic included in update post

- topicname provided as a param in updatePost

re #399

wanlingt committed:
refactor(post, auth): remove PostWithRelations

replace PostWithRelations with Post, as PostWithRelations is redundant

re #399

wanlingt committed:
refactor(post): account for topic in getRelatedPost

- if the post's topic is not null, find related posts belonging to the same topic. Otherwise, use
tags

re #399

wanlingt committed:
refactor(post): remove redundant function from post.service

re #399

LoneRifle commented on client/src/api/types/post.ts:
```diff
@@ -24,6 +25,7 @@ export type GetPostsDto = {
 
 export type CreatePostReqDto = Pick<BasePostDto, 'title' | 'description'> & {
   tagname: string[]
+  topicname: string
```
```suggestion
  topicname: string
```
- Given that topics are optional, we should allow for null topics to be specified
- I've sorely regretted being unable to intervene with casing of parameters back when others were handling the codebase. Could we use camelCasing here please?

LoneRifle commented on client/src/components/QuestionsList/QuestionsList.component.tsx:
```diff
@@ -7,15 +7,15 @@ import {
 interface QuestionsListProps {
   sort: string
-  agency: string
+  agency: number
```
For clarity..
```suggestion
  agencyId: number
```

LoneRifle commented on client/src/components/QuestionsList/QuestionsList.component.tsx:
```diff
@@ -25,35 +25,36 @@ const QuestionsList = ({
         queryKey: [
           LIST_POSTS_QUERY_KEY,
-          { sort, agencyAndTags, page, pageSize },
+          { sort, agency, tags, topics, page, pageSize },
```
Similarly..
```suggestion
          { sort, agencyId, tags, topics, page, pageSize },
```

LoneRifle commented on client/src/components/QuestionsList/QuestionsList.component.tsx:
```diff
@@ -25,35 +25,36 @@ const QuestionsList = ({
+          { sort, agency, tags, topics, page, pageSize },
         ],
-        queryFn: () => listPosts(sort, agencyAndTags, page, pageSize),
+        queryFn: () => listPosts(sort, agency, tags, topics, page, pageSize),
```
and also..
```suggestion
        queryFn: () => listPosts(sort, agencyId, tags, topics, page, pageSize),
```

LoneRifle commented on client/src/pages/EditForm/EditForm.component.tsx:
```diff
@@ -69,6 +69,7 @@ const EditForm = ({ match }: EditFormProps): JSX.Element => {
       description: data.postData.description,
       title: data.postData.title,
       tagname: data.tags,
+      topicname: '', //TODO: update when topics selection is introduced in frontend
```
If the client type supports using null, you should likewise use null
```suggestion
      topicName: null, //TODO: update when topics selection is introduced in frontend
```

LoneRifle commented on server/src/modules/post/post.service.ts:
```diff
@@ -168,127 +235,190 @@ export class PostService {
     size,
   }: {
     sort: SortType
+    agency: number
```
```suggestion
    agencyId: number
```
and elsewhere

LoneRifle commented on server/src/modules/post/post.routes.ts:
```diff
@@ -37,30 +37,12 @@ export const routePosts = ({
     [
       query('withAnswers').isBoolean().toBoolean(),
       query('sort').isIn(Object.values(SortType)),
-      query('tags').customSanitizer((value) => {
-        if (!value) {
-          return undefined
-        }
-        /*
-        Transform req.query.tags into a string array of tags
-        Accept a mixture of query param formats
-        1. comma separated eg. tags=ema,ogp
-        2. repeated query param eg. tags=ema&tags=ogp
-        
-        eg. if query params are tags=parking,parking-activate&tags=ogp
-        req.query.tags = ['parking, parking-activate', 'ogp'] as body-parser accept query param format 2
-        this custom santiser transforms req.query.tags into ['parking', 'parking-activate', 'ogp']
-        */
-        const tagsArray: string[] = [].concat(value)
-        const nestedTags = tagsArray.map((tags) => tags.split(','))
-        return ([] as string[]).concat(...nestedTags)
-      }),
```
wait, why was this removed? It's still needed to validate and sanitise tags. If anything you would need to implement similar for topics too.

wanlingt replied:
The splitting of tags/topics is now done in post.service - previously there was an inconsistency (for listPosts it was done in post.service, but for listAnswerablePosts it was done in post.routes, so I moved it). Is it better if the splitting of tags/topics is done in post.routes though?

wanlingt replied:
ok ignore my reply! just saw your comment below - will move to post.routes

LoneRifle commented on server/src/modules/post/post.service.ts:
```diff
@@ -84,50 +96,105 @@ export class PostService {
+              attributes: [],
+              required: false,
+              where: { id: post.topicId },
+            },
```
You can avoid the join simply by matching on `Post.topicId` in `where`

LoneRifle commented on server/src/modules/post/post.service.ts:
```diff
@@ -84,50 +96,105 @@ export class PostService {
+                'relatedTopics',
+              ],
+              [Sequelize.fn('COUNT', Sequelize.col('tags.id')), 'relatedTags'],
+            ],
```
This would look a little complicated, but to avoid the double `Post.findAll` call, you could specify the desired attributes and order in a separate variable, then spread into the Post, ie,
```ts
const findOptions = post.topicId
  ? { attributes: { include: [...] }, order: [...] }
  : ..... 
```

Alternatively, specify the original findOptions as if there is no topic, and if there is indeed a topic, change `attributes.include` and `order` accordingly.

Then spread into `Post.findAll`:
```ts
const relatedPosts = await this.Post.findAll({
  // ...
  ...findOptions,
  // ......
})
```

LoneRifle commented on server/src/modules/post/post.service.ts:
```diff
@@ -168,127 +235,190 @@ export class PostService {
+
+    let parentAndChildTopics: Topic[] = topics
+
+    if (topics) {
+      let currGenChildTopics = agencyTopics.filter(
+        (topic) =>
+          !!topic.parentId && existingTopicIds.includes(topic.parentId),
+      )
+      parentAndChildTopics = parentAndChildTopics.concat(currGenChildTopics)
+
+      while (currGenChildTopics.length) {
+        const currGenChildTopicIds = currGenChildTopics.map((topic) => topic.id)
+        const nextGenChildTopics = agencyTopics.filter(
+          (topic) =>
+            !!topic.parentId && currGenChildTopicIds.includes(topic.parentId),
+        )
+        parentAndChildTopics = parentAndChildTopics.concat(nextGenChildTopics)
+        currGenChildTopics = nextGenChildTopics
+      }
+    }
```
How does this return a tree?

wanlingt replied:
it doesn't, I didn't think a tree structure was necessary as I just wanted a list of relevant topicIds for the where query in listPosts

LoneRifle commented on server/src/modules/post/post.service.ts:
```diff
@@ -168,127 +235,190 @@ export class PostService {
+    let tagsUnchecked: string[] = []
     if (tags) {
-      tags_unchecked = tags.split(',')
+      tagsUnchecked = tags.split(',')
```
Generally, the service should not be left to handle the parsing of `tags` or `topics` (and in fact, this should already be done by the validation logic in post.routes)

LoneRifle reviewed (changes requested):
quite a bit to get through, so take your time. Ask me to come along to pair if you need.

wanlingt committed:
refactor(post): refactor updatePost to throw valid errors

- updated post.service tests for updatePost
- added null to topicname and tagname tyles
- throws
error if at least one valid tag or topic is not given

re #399

wanlingt committed:
refactor(topic): change topicname to topicName, add null

re #399

wanlingt committed:
refactor(post): change agency to agencyId

- for clarity

re #399

wanlingt committed:
refactor(post): change topicName to topicId in create and update post

- topicId is more specific

re #399

wanlingt committed:
refactor(post): shift splitting of tags/topics from post.service to post.routes

- listPosts and listAnswerablePosts

re #399

LoneRifle commented on server/src/modules/post/post.routes.ts:
```diff
@@ -20,7 +20,48 @@ export const routePosts = ({
+        const nestedTopics = topicsArray.map((topics) => topics.split(','))
+        return ([] as string[]).concat(...nestedTopics)
+      }),
+    ],
```
apply DRY here and have the sanitisers held in a separate array, `sanitizeTagsAndTopics` or similar, then refer to them for both the answerable and public posts endpoints

wanlingt committed:
refactor(post.service): refactor getRelatedPosts seq query

- specify attributes and order in a separate variable findOptions

re #399

wanlingt committed:
fix: resolve merge conflict

- conflicts occured in post.service.ts and web.controller.ts

re #399

wanlingt committed:
test(web): modified unit test for web.controller.spec.ts as parameters were changed in listPosts

re #399

LoneRifle reviewed (approved):
lgtm, good work!

wanlingt committed:
fix(post): fixed createPost error thrown on invalid topic

re #399
